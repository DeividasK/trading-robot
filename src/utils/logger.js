import chalk from "chalk";
import type { Chalk } from "chalk";
import { omit } from "lodash";
import PrettyError from "pretty-error";

const pe = new PrettyError();

const isTesting = process.env.NODE_ENV === "test";

export function log(message: string) {
  if (!isTesting) {
    console.log(message);
  }
}

export function logError(error: any) {
  console.log(pe.render(error));
}

function getSignalColour(signal): "green" | "red" | "yellow" {
  switch (true) {
    case signal === "buy":
      return "green";
    case signal === "sell":
      return "red";
    default:
      return "yellow";
  }
}

export function logTradeRecommendation(
  tradeRecommendation: TradeRecommendation,
  logger: string => void = console.log,
): string {
  const signalColour = getSignalColour(tradeRecommendation.signal);
  const report: Chalk = chalk.keyword(signalColour);

  const message = `------------
SIGNAL: ${report(tradeRecommendation.signal.toUpperCase())}

REASONS:
${tradeRecommendation.reasons.join("\n")}
${
    tradeRecommendation.conditions !== undefined
      ? `
CONDITIONS:
${Object.entries(tradeRecommendation.conditions)
          .map(entry => `${entry[0]}: ${String(entry[1])}`)
          .join("\n")}
------------`
      : `------------`
  }`;

  logger(message);

  return message;
}
