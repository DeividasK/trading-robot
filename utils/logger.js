import PrettyError from "pretty-error";
const pe = new PrettyError();

export function logError(error) {
  console.log(pe.render(error));
}
