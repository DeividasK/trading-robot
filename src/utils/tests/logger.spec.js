import { logTradeRecommendation } from "../logger";

describe("logTradeRecommendation", () => {
  it("should log 'hold' recommendation without conditions", () => {
    const holdRecommendation: HoldRecommendation = {
      signal: "hold",
      reasons: [
        `Couldn't determine a trade signal based on the existing trends. The trends are as follows:
- Short trend - rising
- Medium trend - rising
- Long trend - falling`,
      ],
    };

    expect(
      logTradeRecommendation(holdRecommendation, jest.fn()),
    ).toMatchSnapshot();
  });

  it("should log 'create order' recommendation with conditions", () => {
    const createOrderRecommendation: CreateOrderRecommendation = {
      signal: "buy",
      reasons: [
        "Fast moving average is 0.89718 with a falling trend",
        "Slow moving average is 0.89682 with a falling trend",
        "Overall trend is falling with a 0.89688 average",
      ],
      conditions: {
        isOpen: false,
        price: 0.89682,
        stopLoss: 0.8918,
      },
    };

    expect(
      logTradeRecommendation(createOrderRecommendation, jest.fn()),
    ).toMatchSnapshot();
  });
});
