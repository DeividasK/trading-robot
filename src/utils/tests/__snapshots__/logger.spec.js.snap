// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logTradeRecommendation should log 'create order' recommendation with conditions 1`] = `
"------------
SIGNAL: [38;2;0;128;0mBUY[39m

REASONS:
Fast moving average is 0.89718 with a falling trend
Slow moving average is 0.89682 with a falling trend
Overall trend is falling with a 0.89688 average

CONDITIONS:
isOpen: false
price: 0.89682
stopLoss: 0.8918
------------"
`;

exports[`logTradeRecommendation should log 'hold' recommendation without conditions 1`] = `
"------------
SIGNAL: [38;2;255;255;0mHOLD[39m

REASONS:
Couldn't determine a trade signal based on the existing trends. The trends are as follows:
- Short trend - rising
- Medium trend - rising
- Long trend - falling
------------"
`;
